File: python/code/9-ppg-signal.py
import numpy as np
import matplotlib.pyplot as plt
from scipy.signal import find_peaks

# Sampling frequency
fs = 100

# Duration
duration = 10

# Time
t = np.arange(0, duration, 1 / fs)

# -------------------------
# Generate synthetic PPG signal
# -------------------------
heart_rate = 75
frequency = heart_rate / 60

ppg = (
    1
    + 0.5 * np.sin(2 * np.pi * frequency * t)
    + 0.1 * np.sin(4 * np.pi * frequency * t)
)

# Add small noise
np.random.seed(10)
ppg += 0.03 * np.random.randn(len(t))


# -------------------------
# Find systolic peaks
# -------------------------
peaks, _ = find_peaks(
    ppg,
    distance=int(0.5 * fs),
    prominence=0.1
)


# -------------------------
# Find diastolic points
# -------------------------
inverted_ppg = -ppg

diastolic, _ = find_peaks(
    inverted_ppg,
    distance=int(0.5 * fs),
    prominence=0.05
)


# -------------------------
# Calculate heart rate
# -------------------------
if len(peaks) > 1:
    peak_intervals = np.diff(peaks) / fs
    average_interval = np.mean(peak_intervals)

    calculated_heart_rate = 60 / average_interval
else:
    calculated_heart_rate = 0


print("Number of systolic peaks:", len(peaks))
print("Number of diastolic points:", len(diastolic))
print("Heart Rate:", round(calculated_heart_rate, 2), "BPM")


# -------------------------
# Plot
# -------------------------
plt.figure(figsize=(12, 6))

plt.plot(t, ppg, label="PPG Signal")

plt.plot(
    t[peaks],
    ppg[peaks],
    "ro",
    label="Systolic Peaks"
)

plt.plot(
    t[diastolic],
    ppg[diastolic],
    "go",
    label="Diastolic Points"
)

plt.title("PPG Signal Analysis")
plt.xlabel("Time (s)")
plt.ylabel("Amplitude")
plt.legend()
plt.grid(True)

plt.show()

File: matlab/code/9-ppg-signal.m
clc;
clear;
close all;

% Sampling frequency
Fs = 100;

% Duration
duration = 20;

% Time vector
t = 0:1/Fs:duration-1/Fs;

% Heart rate
heart_rate = 75;

% Heart frequency
f_hr = heart_rate / 60;

% Generate synthetic PPG signal
ppg = 0.8 * sin(2*pi*f_hr*t) ...
    + 0.2 * sin(4*pi*f_hr*t) ...
    + 0.05 * randn(size(t));

% Shift signal upward
ppg = ppg + 1;

% Find systolic peaks
min_peak_distance = round(0.5 * Fs);
min_peak_height = 1.3;

[systolic_peaks, systolic_locations] = findpeaks( ...
    ppg, ...
    'MinPeakDistance', min_peak_distance, ...
    'MinPeakHeight', min_peak_height);

% Find diastolic points
[diastolic_values, diastolic_locations] = findpeaks( ...
    -ppg, ...
    'MinPeakDistance', min_peak_distance);

% Convert negative values back
diastolic_values = -diastolic_values;

% Calculate heart rate
number_of_beats = length(systolic_locations);

measured_hr = ...
    number_of_beats / duration * 60;

% Plot PPG
figure;

plot(t, ppg, 'LineWidth', 1);
hold on;

% Systolic peaks
plot( ...
    t(systolic_locations), ...
    systolic_peaks, ...
    'ro', ...
    'MarkerFaceColor', 'r');

% Diastolic points
plot( ...
    t(diastolic_locations), ...
    diastolic_values, ...
    'go', ...
    'MarkerFaceColor', 'g');

grid on;

xlabel('Time (s)');
ylabel('Amplitude');

title('PPG Signal with Systolic and Diastolic Points');

legend( ...
    'PPG Signal', ...
    'Systolic Peaks', ...
    'Diastolic Points');

hold off;

% Display heart rate
fprintf('Detected Beats = %d\n', number_of_beats);
fprintf('Estimated Heart Rate = %.2f BPM\n', measured_hr);